File: backend/app/routers/analysis.dart
import 'dart:async';
import 'dart:io';
import 'package:http/http.dart' as http;
import 'package:path/path.dart' as path;
import 'dart:convert'; // JSONデコードのために追加

/// オーディオ分析サービス全般で発生する可能性のある例外の基底クラス
class AudioAnalysisException implements Exception {
  final String message;
  final dynamic originalError;
  AudioAnalysisException(this.message, [this.originalError]);

  @override
  String toString() {
    if (originalError != null) {
      return 'AudioAnalysisException: $message ($originalError)';
    }
    return 'AudioAnalysisException: $message';
  }
}

/// オーディオファイルのバリデーションエラーを示す例外
class AudioValidationException extends AudioAnalysisException {
  AudioValidationException(String message) : super(message);
}

/// Kaldiサービスとの通信中に発生したエラーを示す例外
class KaldiServiceException extends AudioAnalysisException {
  final int? statusCode; 

  KaldiServiceException(String message, {this.statusCode, dynamic originalError})
      : super(message, originalError);

  @override
  String toString() {
    String statusInfo = statusCode != null ? ' (Status: $statusCode)' : '';
    return 'KaldiServiceException: $message$statusInfo';
  }
}

class AnalysisResult {
  final bool success;
  final double? score;
  final String? feedback;
  final AnalysisDetails? details;
  final String? errorMessage;

  AnalysisResult({
    required this.success,
    this.score,
    this.feedback,
    this.details,
    this.errorMessage,
  });

  factory AnalysisResult.fromJson(Map<String, dynamic> json) {
    return AnalysisResult(
      success: json['success'] as bool,
      score: (json['score'] as num?)?.toDouble(), // null許容に対応
      feedback: json['feedback'] as String?,
      details: json['details'] != null
          ? AnalysisDetails.fromJson(json['details'])
          : null,
      errorMessage: json['error'] as String?,
    );
  }

  factory AnalysisResult.error(String message, {dynamic originalError}) {
    // エラー発生時に元のエラー情報を渡せるように改善
    return AnalysisResult(
      success: false,
      errorMessage: message + (originalError != null ? ' ($originalError)' : ''),
    );
  }

  @override
  String toString() {
    if (success) {
      return 'AnalysisResult(success: true, score: $score, feedback: "$feedback", details: $details)';
    } else {
      return 'AnalysisResult(success: false, errorMessage: "$errorMessage")';
    }
  }
}

class AnalysisDetails {
  final double pronunciation;
  final double intonation;
  final double rhythm;

  const AnalysisDetails({
    required this.pronunciation,
    required this.intonation,
    required this.rhythm,
  });

  factory AnalysisDetails.fromJson(Map<String, dynamic> json) {
    try {
      return AnalysisDetails(
        pronunciation: (json['pronunciation'] as num).toDouble(),
        intonation: (json['intonation'] as num).toDouble(),
        rhythm: (json['rhythm'] as num).toDouble(),
      );
    } catch (e) {
      throw FormatException('Invalid JSON format for AnalysisDetails: $e');
    }
  }

  Map<String, dynamic> toJson() => {
        'pronunciation': pronunciation,
        'intonation': intonation,
        'rhythm': rhythm,
      };

  bool isValid() {
    return pronunciation >= 0 &&
        pronunciation <= 100 &&
        intonation >= 0 &&
        intonation <= 100 &&
        rhythm >= 0 &&
        rhythm <= 100;
  }

  @override
  String toString() =>
      'AnalysisDetails(pronunciation: $pronunciation, intonation: $intonation, rhythm: $rhythm)';
}

class AnalyzeService {
  // 定数をより具体的に
  static const String kaldiServiceUrl = String.fromEnvironment(
    'KALDI_SERVICE_URL',
    defaultValue: 'http://localhost:8080/analyze'
  );
  
  // キャッシュの追加
  final _cache = <String, AnalysisResult>{};
  
  // リトライ機能の追加
 Future<AnalysisResult> analyzeAudioWithRetry(
    File audioFile, {
    int maxRetries = 3,
    Duration retryDelay = const Duration(seconds: 1),
  }) async {
    for (var i = 0; i < maxRetries; i++) {
      try {
        return await analyzeAudio(audioFile);
      } on KaldiServiceException catch (e) {
        if (i == maxRetries - 1) rethrow;
        await Future.delayed(retryDelay * (i + 1));
      }
    }
    throw KaldiServiceException('Max retries exceeded');
  }
  
    // タイムアウトの設定
    Future<void> preloadModels() async {
  }

  // メモリ解放
  void dispose() {
    _cache.clear();
  }
}

  /// メッセージをログに出力します。
  void _log(String message, {bool isError = false}) {
    if (isError) {
      // 実際には logger パッケージなどを利用
      print('ERROR: AnalyzeService - $message');
    } else {
      print('INFO: AnalyzeService - $message');
    }
  }

  /// オーディオファイルを分析し、AnalysisResultを返します。
  Future<AnalysisResult> analyzeAudio(File audioFile) async {
    _log('Starting audio analysis for: ${audioFile.path}');
    try {
      // ファイルバリデーション
      await _validateAudioFile(audioFile);

      // Kaldiサービスにリクエスト送信
      final Map<String, dynamic> rawResponse =
          await _sendToKaldiService(audioFile);

      // 結果の解析と加工
      return _processAnalysisResult(rawResponse);
    } on AudioValidationException catch (e) {
      _log('Audio validation error: ${e.message}', isError: true);
      return AnalysisResult.error('Invalid audio file: ${e.message}', originalError: e);
    } on KaldiServiceException catch (e) {
      _log('Kaldi service communication error: ${e.message}', isError: true);
      return AnalysisResult.error('Failed to communicate with analysis service: ${e.message}', originalError: e);
    } on TimeoutException catch (e) {
      _log('Request timed out: $e', isError: true);
      return AnalysisResult.error('Analysis request timed out. Please try again.', originalError: e);
    } catch (e) {
      _log('An unexpected error occurred: $e', isError: true);
      return AnalysisResult.error('An unexpected error occurred during analysis: $e', originalError: e);
    }
  }

  /// オーディオファイルのバリデーションを行います。
  /// 無効な場合は [AudioValidationException] をスローします。
  Future<void> _validateAudioFile(File file) async {
    final extension = path.extension(file.path).toLowerCase();
    _log('Validating file: ${file.path}, extension: $extension');

    if (!['.wav', '.mp3'].contains(extension)) {
      throw AudioValidationException('Unsupported audio file format: $extension. Only .wav and .mp3 are supported.');
    }

    // ファイルサイズのチェック
    final fileSize = await file.length(); 
    if (fileSize > maxFileSizeMB * 1024 * 1024) {
      throw AudioValidationException(
          'Audio file size exceeds the limit of ${maxFileSizeMB}MB.');
    }

    // MIMEタイプバリデーションのプレースホルダー
    // 実際のアプリケーションでは、file_pickerなどからMIMEタイプを取得して検証することが多い
    // String? mimeType = await file_picker_package.getMimeType(file.path);
    // if (mimeType != null && !['audio/wav', 'audio/mpeg'].contains(mimeType)) {
    //   throw AudioValidationException('Invalid MIME type for audio file.');
    // }

    _log('Audio file validation successful.');
  }

  /// Kaldiサービスにオーディオファイルを送信し、その生のレスポンスを返します。
  /// エラーが発生した場合は [KaldiServiceException] または [TimeoutException] をスローします。
  Future<Map<String, dynamic>> _sendToKaldiService(File audioFile) async {
    _log('Sending audio file to Kaldi service: ${audioFile.path}');
    try {
      var request = http.MultipartRequest('POST', Uri.parse(kaldiServiceUrl));
      request.files.add(
        await http.MultipartFile.fromPath('audio_file', audioFile.path),
      );

      final streamedResponse = await request.send().timeout(timeoutDuration);
      final response = await http.Response.fromStream(streamedResponse);

      _log('Received response from Kaldi service with status: ${response.statusCode}');

      if (response.statusCode != 200) {
        // エラーレスポンスボディをログに記録
        _log('Kaldi service returned error: ${response.body}', isError: true);
        throw KaldiServiceException(
            'Kaldi service returned an error. Status: ${response.statusCode}',
            statusCode: response.statusCode,
            originalError: response.body);
      }

      // レスポンスボディがJSON形式であることを期待
      try {
        return json.decode(response.body) as Map<String, dynamic>;
      } on FormatException catch (e) {
        _log('Failed to parse Kaldi service response as JSON: ${response.body}', isError: true);
        throw KaldiServiceException(
            'Failed to parse analysis result from service.',
            originalError: e);
      }
    } on SocketException catch (e) {
      throw KaldiServiceException('Network error communicating with Kaldi service: $e', originalError: e);
    } on TimeoutException catch (e) {
      throw TimeoutException('Request to Kaldi service timed out after ${timeoutDuration.inSeconds} seconds.', e.duration);
    } catch (e) {
      throw KaldiServiceException('Failed to communicate with Kaldi service: $e', originalError: e);
    }
  }

  /// Kaldiサービスからの生のレスポンスを解析し、[AnalysisResult] オブジェクトに変換します。
  AnalysisResult _processAnalysisResult(Map<String, dynamic> rawResponse) {
    _log('Processing raw analysis result.');
    try {
      // Kaldiからの生のレスポンス構造に応じて、AnalysisResult.fromJson に渡すデータを整形
      // 例: Kaldiが直接 'score', 'feedback', 'details' を返す場合
      // もしKaldiからのレスポンスが異なる構造の場合、ここで変換ロジックを追加する必要があります。
      // 例: rawResponse['result']['final_score'] のようなネストされた構造
      // 今回は、rawResponseが直接 AnalysisResult.fromJson に適合する形だと仮定します。
      final analysisResult = AnalysisResult.fromJson(rawResponse);

      // ここでさらにビジネスロジックに基づいてスコアやフィードバックを調整することも可能
      // 例: _calculateScore(rawResponse) などのメソッドは、JSONから取得した値を元に調整する場合に使用
      if (!analysisResult.success) {
        _log('Analysis result indicates failure: ${analysisResult.errorMessage}', isError: true);
        return AnalysisResult.error('Analysis was not successful: ${analysisResult.errorMessage}');
      }

      if (analysisResult.details != null && !analysisResult.details!.isValid()) {
         _log('Analysis details are out of valid range: ${analysisResult.details}', isError: true);
         // スコアや詳細の範囲外チェックが真であればエラーとして処理
         return AnalysisResult.error('Analysis details contain invalid values.');
      }

      _log('Analysis result processed successfully.');
      return analysisResult;
    } on FormatException catch (e) {
      _log('Failed to parse analysis result from Kaldi service: $e', isError: true);
      return AnalysisResult.error('Invalid analysis result format from service: $e', originalError: e);
    } catch (e) {
      _log('Error processing analysis result: $e', isError: true);
      return AnalysisResult.error('Error processing analysis result: $e', originalError: e);
    }
  }

File: backend/app/routers/analysis.py
import asyncio
import aiohttp
import os
import json
from pathlib import Path
from typing import Optional, Dict, Any, Union
import magic
from dataclasses import dataclass
import logging

# ログ設定
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class AudioAnalysisException(Exception):
    """オーディオ分析サービス全般で発生する可能性のある例外の基底クラス"""
    
    def __init__(self, message: str, original_error: Optional[Exception] = None):
        self.message = message
        self.original_error = original_error
        super().__init__(self.message)
    
    def __str__(self) -> str:
        if self.original_error is not None:
            return f'AudioAnalysisException: {self.message} ({self.original_error})'
        return f'AudioAnalysisException: {self.message}'


class AudioValidationException(AudioAnalysisException):
    """オーディオファイルのバリデーションエラーを示す例外"""
    
    def __init__(self, message: str):
        super().__init__(message)


class KaldiServiceException(AudioAnalysisException):
    """Kaldiサービスとの通信中に発生したエラーを示す例外"""
    
    def __init__(self, message: str, status_code: Optional[int] = None, original_error: Optional[Exception] = None):
        self.status_code = status_code
        super().__init__(message, original_error)
    
    def __str__(self) -> str:
        status_info = f' (Status: {self.status_code})' if self.status_code is not None else ''
        return f'KaldiServiceException: {self.message}{status_info}'


@dataclass
class AnalysisDetails:
    """分析の詳細情報を保持するクラス"""
    pronunciation: float
    intonation: float
    rhythm: float
    
    @classmethod
    def from_json(cls, json_data: Dict[str, Any]) -> 'AnalysisDetails':
        try:
            return cls(
                pronunciation=float(json_data['pronunciation']),
                intonation=float(json_data['intonation']),
                rhythm=float(json_data['rhythm'])
            )
        except (KeyError, ValueError, TypeError) as e:
            raise ValueError(f'Invalid JSON format for AnalysisDetails: {e}')
    
    def to_json(self) -> Dict[str, float]:
        return {
            'pronunciation': self.pronunciation,
            'intonation': self.intonation,
            'rhythm': self.rhythm
        }
    
    def is_valid(self) -> bool:
        return (0 <= self.pronunciation <= 100 and
                0 <= self.intonation <= 100 and
                0 <= self.rhythm <= 100)
    
    def __str__(self) -> str:
        return f'AnalysisDetails(pronunciation: {self.pronunciation}, intonation: {self.intonation}, rhythm: {self.rhythm})'


@dataclass
class AnalysisResult:
    """分析結果を保持するクラス"""
    success: bool
    score: Optional[float] = None
    feedback: Optional[str] = None
    details: Optional[AnalysisDetails] = None
    error_message: Optional[str] = None
    
    @classmethod
    def from_json(cls, json_data: Dict[str, Any]) -> 'AnalysisResult':
        details = None
        if json_data.get('details') is not None:
            details = AnalysisDetails.from_json(json_data['details'])
        
        return cls(
            success=bool(json_data['success']),
            score=float(json_data['score']) if json_data.get('score') is not None else None,
            feedback=json_data.get('feedback'),
            details=details,
            error_message=json_data.get('error')
        )
    
    @classmethod
    def error(cls, message: str, original_error: Optional[Exception] = None) -> 'AnalysisResult':
        """エラー発生時に元のエラー情報を渡せるように改善"""
        error_msg = message + (f' ({original_error})' if original_error is not None else '')
        return cls(
            success=False,
            error_message=error_msg
        )
    
    def __str__(self) -> str:
        if self.success:
            return f'AnalysisResult(success: True, score: {self.score}, feedback: "{self.feedback}", details: {self.details})'
        else:
            return f'AnalysisResult(success: False, error_message: "{self.error_message}")'


class AnalyzeService:
    """オーディオ分析サービスのメインクラス"""
    
    # 定数をより具体的に
    KALDI_SERVICE_URL = os.environ.get('KALDI_SERVICE_URL', 'http://localhost:8080/analyze')
    MAX_FILE_SIZE_MB = 50  # デフォルト値として設定
    TIMEOUT_DURATION = 120  # 秒
    
    def __init__(self):
        # キャッシュの追加
        self._cache: Dict[str, AnalysisResult] = {}
    
    def _log(self, message: str, is_error: bool = False) -> None:
        """メッセージをログに出力します。"""
        if is_error:
            logger.error(f'AnalyzeService - {message}')
        else:
            logger.info(f'AnalyzeService - {message}')
    
    async def analyze_audio_with_retry(self, audio_file: Union[str, Path], 
                                     max_retries: int = 3, 
                                     retry_delay: float = 1.0) -> AnalysisResult:
        """リトライ機能付きでオーディオを分析"""
        for i in range(max_retries):
            try:
                return await self.analyze_audio(audio_file)
            except KaldiServiceException as e:
                if i == max_retries - 1:
                    raise
                await asyncio.sleep(retry_delay * (i + 1))
        
        raise KaldiServiceException('Max retries exceeded')
    
    async def preload_models(self) -> None:
        """モデルのプリロード（プレースホルダー）"""
        pass
    
    def dispose(self) -> None:
        """メモリ解放"""
        self._cache.clear()
    
    async def analyze_audio(self, audio_file: Union[str, Path]) -> AnalysisResult:
        """オーディオファイルを分析し、AnalysisResultを返します。"""
        audio_path = Path(audio_file)
        self._log(f'Starting audio analysis for: {audio_path}')
        
        try:
            # ファイルバリデーション
            await self._validate_audio_file(audio_path)
            
            # Kaldiサービスにリクエスト送信
            raw_response = await self._send_to_kaldi_service(audio_path)
            
            # 結果の解析と加工
            return self._process_analysis_result(raw_response)
            
        except AudioValidationException as e:
            self._log(f'Audio validation error: {e.message}', is_error=True)
            return AnalysisResult.error(f'Invalid audio file: {e.message}', original_error=e)
            
        except KaldiServiceException as e:
            self._log(f'Kaldi service communication error: {e.message}', is_error=True)
            return AnalysisResult.error(f'Failed to communicate with analysis service: {e.message}', original_error=e)
            
        except asyncio.TimeoutError as e:
            self._log(f'Request timed out: {e}', is_error=True)
            return AnalysisResult.error('Analysis request timed out. Please try again.', original_error=e)
            
        except Exception as e:
            self._log(f'An unexpected error occurred: {e}', is_error=True)
            return AnalysisResult.error(f'An unexpected error occurred during analysis: {e}', original_error=e)
    
    async def _validate_audio_file(self, file_path: Path) -> None:
        """オーディオファイルのバリデーションを行います。
        無効な場合は AudioValidationException をスローします。"""
        
        extension = file_path.suffix.lower()
        self._log(f'Validating file: {file_path}, extension: {extension}')
        
        if extension not in ['.wav', '.mp3']:
            raise AudioValidationException(
                f'Unsupported audio file format: {extension}. Only .wav and .mp3 are supported.'
            )
        
        # ファイルサイズのチェック
        if not file_path.exists():
            raise AudioValidationException(f'Audio file does not exist: {file_path}')
        
        file_size = file_path.stat().st_size
        if file_size > self.MAX_FILE_SIZE_MB * 1024 * 1024:
            raise AudioValidationException(
                f'Audio file size exceeds the limit of {self.MAX_FILE_SIZE_MB}MB.'
            )
        
        mime_type = MIME_MAGIC.from_file(str(file_path))
        if mime_type not in ALLOWED_AUDIO_TYPES:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"不正なファイル形式です: {mime_type}"
            )
        
        self._log('ファイルのバリデーションが成功しました。')
    
    async def _send_to_kaldi_service(self, audio_file: Path) -> Dict[str, Any]:
        """Kaldiサービスにオーディオファイルを送信し、その生のレスポンスを返します。
        エラーが発生した場合は KaldiServiceException または TimeoutError をスローします。"""
        
        self._log(f'Sending audio file to Kaldi service: {audio_file}')
        
        try:
            async with aiohttp.ClientSession(timeout=aiohttp.ClientTimeout(total=self.TIMEOUT_DURATION)) as session:
                with open(audio_file, 'rb') as f:
                    data = aiohttp.FormData()
                    data.add_field('audio_file', f, filename=audio_file.name)
                    
                    async with session.post(self.KALDI_SERVICE_URL, data=data) as response:
                        self._log(f'Received response from Kaldi service with status: {response.status}')
                        
                        if response.status != 200:
                            error_body = await response.text()
                            self._log(f'Kaldi service returned error: {error_body}', is_error=True)
                            raise KaldiServiceException(
                                f'Kaldi service returned an error. Status: {response.status}',
                                status_code=response.status,
                                original_error=error_body
                            )
                        
                        # レスポンスボディがJSON形式であることを期待
                        try:
                            return await response.json()
                        except (json.JSONDecodeError, aiohttp.ContentTypeError) as e:
                            error_text = await response.text()
                            self._log(f'Failed to parse Kaldi service response as JSON: {error_text}', is_error=True)
                            raise KaldiServiceException(
                                'Failed to parse analysis result from service.',
                                original_error=e
                            )
        
        except aiohttp.ClientError as e:
            raise KaldiServiceException(f'Network error communicating with Kaldi service: {e}', original_error=e)
        
        except asyncio.TimeoutError as e:
            raise asyncio.TimeoutError(f'Request to Kaldi service timed out after {self.TIMEOUT_DURATION} seconds.')
        
        except Exception as e:
            raise KaldiServiceException(f'Failed to communicate with Kaldi service: {e}', original_error=e)
    
    def _process_analysis_result(self, raw_response: Dict[str, Any]) -> AnalysisResult:
        """Kaldiサービスからの生のレスポンスを解析し、AnalysisResult オブジェクトに変換します。"""
        
        self._log('Processing raw analysis result.')
        
        try:
            # Kaldiからの生のレスポンス構造に応じて、AnalysisResult.from_json に渡すデータを整形
            # 例: Kaldiが直接 'score', 'feedback', 'details' を返す場合
            # もしKaldiからのレスポンスが異なる構造の場合、ここで変換ロジックを追加する必要があります。
            # 例: raw_response['result']['final_score'] のようなネストされた構造
            # 今回は、raw_responseが直接 AnalysisResult.from_json に適合する形だと仮定します。
            
            analysis_result = AnalysisResult.from_json(raw_response)
            
            # ここでさらにビジネスロジックに基づいてスコアやフィードバックを調整することも可能
            # 例: _calculate_score(raw_response) などのメソッドは、JSONから取得した値を元に調整する場合に使用
            
            if not analysis_result.success:
                self._log(f'Analysis result indicates failure: {analysis_result.error_message}', is_error=True)
                return AnalysisResult.error(f'Analysis was not successful: {analysis_result.error_message}')
            
            if analysis_result.details is not None and not analysis_result.details.is_valid():
                self._log(f'Analysis details are out of valid range: {analysis_result.details}', is_error=True)
                # スコアや詳細の範囲外チェックが真であればエラーとして処理
                return AnalysisResult.error('Analysis details contain invalid values.')
            
            self._log('Analysis result processed successfully.')
            return analysis_result
            
        except ValueError as e:
            self._log(f'Failed to parse analysis result from Kaldi service: {e}', is_error=True)
            return AnalysisResult.error(f'Invalid analysis result format from service: {e}', original_error=e)
        
        except Exception as e:
            self._log(f'Error processing analysis result: {e}', is_error=True)
            return AnalysisResult.error(f'Error processing analysis result: {e}', original_error=e)


# 使用例
async def main():
    service = AnalyzeService()
    
    try:
        # 通常の分析
        result = await service.analyze_audio("path/to/audio.wav")
        print(result)
        
        # リトライ付き分析
        result_with_retry = await service.analyze_audio_with_retry("path/to/audio.wav", max_retries=5)
        print(result_with_retry)
        
    finally:
        service.dispose()


if __name__ == "__main__":
    asyncio.run(main())